xapian-letor/api/featurevector_internal.h

```diff
@@ -0,0 +1,76 @@
+/* featurevector.cc: The file responsible for transforming the document into the feature space.
+ *
+ * Copyright (C) 2012 Parth Gupta
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#ifndef FEATURE_VECTOR_INTERNAL_H
+#define FEATURE_VECTOR_INTERNAL_H
+
+#include <config.h>
+
+#include <xapian.h>
+
+#include "xapian-letor/featurevector.h"
+#include "xapian-letor/featuremanager.h"
+
+#include <list>
+#include <map>
+
+#include "str.h"
+#include "stringutils.h"
+
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include "safeerrno.h"
+#include "safeunistd.h"
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+
+
+using namespace std;
+
+namespace Xapian{
+
+class FeatureVector::Internal : public Xapian::Internal::intrusive_base
+{
+    friend class FeatureVector;
+    double label;
+    double score;
+    std::map<int,double> fvals;
```

Comment:
I would suggest to use logical names for the features than simple integers e.g. Enums. Usually a good way is to use different ranges for different types such as 100s for query (Q) related, 200s for document (D) related and 300s for the T and D related. That way it would be easier to add more features later.